stellargraph/data/explorer.py

```diff
@@ -303,26 +303,8 @@ def naive_weighted_choices(rs, weights):
     requires a lot of preprocessing (normalized probabilties), and
     does a lot of conversions/checks/preprocessing internally.
     """
-
-    # divide the interval [0, sum(weights)) into len(weights)
-    # subintervals [x_i, x_{i+1}), where the width x_{i+1} - x_i ==
-    # weights[i]
-    subinterval_ends = []
-    running_total = 0
-    for w in weights:
-        if w < 0:
-            raise ValueError("Detected negative weight: {}".format(w))
-        running_total += w
-        subinterval_ends.append(running_total)
-
-    # pick a place in the overall interval
-    x = rs.random() * running_total
-
-    # find the subinterval that contains the place, by looking for the
-    # first subinterval where the end is (strictly) after it
-    for idx, end in enumerate(subinterval_ends):
-        if x < end:
-            break
+    probs = np.cumsum(weights) / weights.sum()
+    idx = np.searchsorted(probs, rs.random(), side="right")
```

Comment:
I think this is mildly suboptimal for a few reasons:

- it does two passes over the `weights` array, which (a) is probably slightly slower than it needs to be, and (b) may result in a cumulative sum > the final sum, so `probs[-1]` > 1.
- a division is much slower than a multiplication
- doing O(len(weights)) operations is slower than a single operation

I think we can address each of these, e.g.: an "ablation" study:

```python
import numpy as np
N = 10000
weights = np.random.rand(N).astype(np.float32)
rs = np.random.RandomState()

def sum_div():
    probs = np.cumsum(weights) / weights.sum()
    idx = np.searchsorted(probs, rs.random(), side="right")
    
def sum_div_inplace():
    probs = np.cumsum(weights) 
    probs /= weights.sum()
    idx = np.searchsorted(probs, rs.random(), side="right")

def last_div_inplace():
    probs = np.cumsum(weights)
    probs /= probs[-1]
    idx = np.searchsorted(probs, rs.random(), side="right")
    
def last_mul():
    probs = np.cumsum(weights)
    idx = np.searchsorted(probs, rs.random() * probs[-1], side="right")

%timeit -n 10000 sum_div()
# 44.2 µs ± 2.3 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

%timeit -n 10000 sum_div_inplace()
# 38.6 µs ± 1.4 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

%timeit -n 10000 last_div_inplace()
# 38.7 µs ± 1.83 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

%timeit -n 10000 last_mul()
# 36.2 µs ± 1.23 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

| N | time `sum_div` | time `last_mul` |
|---|---|---|
| 3 | 11.5 | 9.8 |
| 10 | 11.7 |  9.88 |
| 100 | 11.9 | 9.89 |
| 1000 | 15.6 | 12.4 |
| 10000 | 44.2 | 36.2 |

```suggestion
    probs = np.cumsum(weights)
    idx = np.searchsorted(probs, rs.random() * probs[-1], side="right")
```